### ISSUE TITLE
Support for Webpack Dev Sever "Hot Module Replacement"

### ISSUE THREAD

**Issue by Scott-MacD** [wix/stylable#401]
The stylable webpack plugin doesn't seem to play nicely with webpack's hot module replacement feature.

I've built a simple rendering/update pipeline separated from state, that once a component is updated, webpack recompiles the assets, hot swaps them into the page, and then we re-render the page without forcing a refresh, keeping our current state in tact (similar to the popular redux demos a while back). When updating a stylable stylesheet however, they are not properly detected as a dependency of a component, and therefore not understood by the HMR that it is handled by the HMR accept function, and therefore causes a full page refresh unless I add an exception for each individual stylesheet. Ideally stylable stylesheets should be properly detected as a dependency and allow the page to be re-rendered without forcing a refresh. Alternatively, a flag in the plugin options similar to the `hmr` option in [style-loader](https://github.com/webpack-contrib/style-loader) would work as well.

I've created a test repo with as much of a reduced test case of this setup as I can, while still demonstrating the intended functionality. Modifying anything in the `actions` or `components` folder just updates the page, updating a `.st.css` file causes a full page refresh.

https://github.com/Scott-MacD/stylable-webpack-hmr

**Comment by idoros** [comment]
Thanks for reporting and making the effort of reproducing the case.

This is technical debt that we are aware of and exists in our backlog. We are missing "stylesheet-hot-reload" that will propagate a change to dependent `js` and other `st.css` files. 

Unfortunately I cannot specify when this will be resolved, hopefully we will get to it in the near future.

**Comment by Scott-MacD** [comment]
No problem Ido, thanks for the update.

Although support for hot module reloading is definitely nice to have, I
wouldn’t say it’s critical for development very often. As a work around
until this gets patched, if someone is finding that they’re working on
particular style sheets often, they should be able to manually add them to
the HRM “accepts” parameters to handle the update in a similar manner and
prevent refresh.

**Comment by Scott-MacD** [comment]
Just an update on this, I can't seem to get Webpack HMR working with Stylable at all anymore, even with manually supplying the path to `module.hot.accepts()`. I'm trying to investigate, but not sure where this change occurred that caused an issue (not sure if it's a change in stylable-webpack-plugin, webpack-dev-server, webpack, or something I just broke altogether).

The issue that seems to be occurring is that HMR's parameters seem to require modules relative to the src directory of a project, while the actual generated module seems to be reporting its path as relative to the project root. Since they are different, HMR sees this as an unhandled update, and therefore causes a full page refresh.

eg. Given a project structure something like:
```
src/
   components/
      my-component.js
      my-component.st.css
   index.js
package.json
webpack.config.js
```

and the following update function:
```javascript
// Webpack HMR
if (module && module.hot) {
    module.hot.accept([
        "./components/my-component.js",
        "./components/my-component.st.css"
    ], update);
}
```

`my-component.js` works as expected, while `my-component.st.css` fails.

If anyone has any idea on how to get this to play nicely I'd love to hear it. All dependancies are updated to their latest version at the time of this post, running on node v 10.14.1

**Comment by tomrav** [comment]
I took a quick look at this issue but did not make any meaningful progress with debugging.

If you could share more information about how your project is structured/configured that could help us figure it out. If you'd be able to update your previous example repo to include your HMR setup, that would be even better.

**Comment by Scott-MacD** [comment]
Not a problem, I’ll try and update the example repo tomorrow.

**Comment by Scott-MacD** [comment]
I've updated the repo at https://github.com/Scott-MacD/stylable-webpack-hmr

I've left the master branch as-is, as that is how everything should ideally work (original purpose of this issue), however I've created a separate branch of `work-around` that has a single commit with the work around that I was using previously. It can get a little bit tedious for larger projects, but it worked.

If you pull that branch and run `npm start` it will spin up Webpack dev server with HMR enabled. You can then edit any of the stylesheets, and although they are accepted by Webpack on the initial compile, it detects the wrong file path and forces a full refresh instead of just updating the page.

When editing `button.st.css`, these are the warning messages that you should see logged before the browser reloads the page:
```
[HMR] Cannot apply update. Need to do a full reload!
[HMR] Error: Aborted because ./src/components/button.st.css is not accepted
Update propagation: ./src/components/button.st.css
    at hotApply (http://localhost:3001/main.js:477:30)
    at http://localhost:3001/main.js:315:22
```

If we try to modify the path to include `src/` as part of the path, it should be correctly accepted by HMR, however refuses to compile because it can't find the directory when Webpack is running it's static analysis (it's trying to load `src` from within `src` as a root directory).

If we were to omit having a src directory and instead work directly in the project root, the work around would function correctly (again, not the ideal solution to this, but should work).

As always, thanks for the help and the awesome work you guys are doing.

**Comment by tomrav** [comment]
Thanks for the reproduction, it really helped in debugging this.

I believe we were able to locate the underlying issue, and plan on publishing a fix early next week (which will hopefully make HMR work out of the box for `*.st.css` files).

**Comment by Scott-MacD** [comment]
Thanks Tom,

Glad I was able to help pinpoint the issue.

**Comment by tomrav** [comment]
I just released a new version (`v0.1.22`) of the webpack plugin that includes a flag (`experimentalHMR`) which enables HMR support out of the box.

Once we see this feature as stable we'll flip (or remove) the flag, and offer this behavior as the default one for development mode.

Lastly, I opened a PR with your example project (Scott-MacD/stylable-webpack-hmr#1) to demo the config, as also shown below.

### Config example
```ts
import StylableWebpackPlugin from '@stylable/webpack-plugin'

new StylableWebpackPlugin({
    experimentalHMR: true
})
```

**Comment by Scott-MacD** [comment]
Thanks Stylable team!

I pulled in the PR, and other than having to re-add HMR resolving for my components themselves (this is completely independent of stylable) it seems to work perfectly as expected now. I've pulled the change into master of the test case and will happily leave it there in case it's useful for anyone.

**Comment by Scott-MacD** [comment]
Just realized this is still under the experimental flag and I prematurely closed this. Re-opening.

**Comment by Scott-MacD** [comment]
I've been using this for a while now and it's made development while working with components much more pleasant to work with, however have noticed that when updating the stylesheet it doesn't propagate the changes back to the JS runtime. This means your components on the page don't get updated to reflect changes to the stylesheet. 

One of the cases where this fails to result in the expected behavior, is if you refactor a class to extend another class. As an example If I have a stylesheet similar to the following:
```css
.primary-button {
    appearance: none;
    border-radius: .25em;
    cursor: pointer;
    display: inline-block;
    user-select: none;
    background: blue;
}

.secondary-button {
    appearance: none;
    border-radius: .25em;
    cursor: pointer;
    display: inline-block;
    user-select: none;
    background: green;
}
```

And I refactor out the common styles into a `.button` class using `-st-extends` 
```css
.button {
    appearance: none;
    border-radius: .25em;
    cursor: pointer;
    display: inline-block;
    user-select: none;
}

.primary-button {
    -st-extends: button;
    background: blue;
}

.secondary-button {
    -st-extends: button;
    background: green;
}
```

My components on the page will lose all the `.button` css because the stylesheet will be updated, but my js importing the stylesheet won't know that a change has occurred, and to add the `.button` class to my existing components using the `.primary-button` and `.secondary-button` classes. In order to force the layout to update you need to either make a change to your component JS file, or reload your browser.

Most of the time this isn't a problem, as the styles themselves are updated correctly and injected into the page, and if I'm creating a new class in a stylesheet I will most likely be editing my components' JS file to add that class to the markup which will be picked up by HMR and update the page as expected. The issue is specifically when you make a change that results in a change to the exported JS object, but aren't making any code changes to JS component itself.

**Comment by barak007** [comment]
I am guessing that it's related to the fact that we do not patch the module when accepting it in HMR.

It might also be related to the fact that the DOM should be updated with the new class mapping (re-rendering the whole app).
Or it might also happen because the component caches the mapped classes values.

I need to further investigate what and how to solve this issue, and will report back here.

Thanks for the update!

**Comment by Scott-MacD** [comment]
Not a problem!

I've updated my reduced test case repo to highlight this issue if that helps:
https://github.com/Scott-MacD/stylable-webpack-hmr

If you make a change to `button.st.css` for example, it will updated the style tag injected into the page, and you get a console log in your browser like the following:
```
[HMR] Updated modules:
[HMR]  - ./src/components/button.st.css
[HMR] App is up to date.
```

If I make a change to `button.js` instead, it automatically re-renders the components and you can see that HMR has dedected the dependancy tree:
```
[HMR] Updated modules:
[HMR]  - ./src/components/button.js
[HMR]  - ./src/components/container.js
[HMR]  - ./src/index.js
[HMR] App is up to date.
```

To reproduce the specific issue I was mentioning, you can comment out `-st-extends` on one of the buttons, and you will see the change is not reflected on the page. If this was working as intended I would expect HMR to detect the dependancy tree of the stylesheet change and log the following, as it is a dependancy of button.js:
```
[HMR] Updated modules:
[HMR]  - ./src/components/button.st.css
[HMR]  - ./src/components/button.js
[HMR]  - ./src/components/container.js
[HMR]  - ./src/index.js
[HMR] App is up to date.
```

It doesn't look like the issue is related to the component caching the style values, as the way this is set up the component automatically re-renders itself any time it detects an update. Also due to the way webpack works in a dev environment, it restructures all your imports as getters so that it always references your latest code any time it compiles.

**Comment by tomrav** [comment]
@Scott-MacD, it's been a while, but we're curious to see if this issue has been resolved by our added flag, or is something still missing?

**Comment by Scott-MacD** [comment]
Hi @tomrav, thanks for following up.

It looks like at some point during the refactoring of the mono repo some formatting clean up broke this flag altogether. There is a comment that got merged with the actual if statement that enables this functionality, therefore preventing that code from getting executed at all.

See: https://github.com/wix/stylable/blob/master/packages/webpack-plugin/src/stylable-generator.ts#L126

Once I correct that statement by removing the comment, the functionality is still the same as my last post. The way the HMR feature works is definitely an improvement over not having it at all, but it still has issues with not updating the JS import if I make changes to the .st.css file (see previous post for more detail).

I've updated the dependancies in my test-repo if that helps:
https://github.com/Scott-MacD/stylable-webpack-hmr

**Comment by tomrav** [comment]
Just merged #858 which should restore the flag functionality, hope to release tomorrow.

Thanks for the catch!

**Comment by Scott-MacD** [comment]
Not a problem, I appreciate the work you guys have done and am happy to help where I can. I'm not super familiar with how either webpack plugins work, or how the entirety of build process of stylable works behind the scenes, but I'll try to find some time to see if I can figure out a way to signal to the HMR that the JS modules need to be hot updated, not just the generated css.
